### PR TITLE
ENH: stats: Additional distributions for scipy.stats commonly used in High Energy Physics

### PR DESCRIPTION
**Pull request by thomaskeck**
A common task in High Energy Physics (HEP) is fitting recorded data with parametrized distributions to extract signal and background components.

I personally started using scipy instead of the traditionally used ROOT framework to do so.
I implemented several distributions which are commonly used in HEP.
However the usage / usefulness is not restricted to HEP.

This pull-requests contains the implementation and unittests for:

**1. The Argus distribution**
One can think of a beta-distribution multiplied by a gaussian distribution.
https://en.wikipedia.org/wiki/ARGUS_distribution

**2. The Crystalball distribution**
A Gaussian with a one-sided power-law tail
https://en.wikipedia.org/wiki/Crystal_Ball_function

**3. A distribution generator "template_gen"** to create a scipy distribution out of a given histogram.
This is often used to incorporate distributions which can be measured or simulated. The techniques is sometimes called "template fit". For reference: ROOT implements it using a class called RooHistPdf
https://root.cern.ch/doc/master/classRooHistPdf.html

**4. A mixture distribution generator**  "mixture_gen" to create a superposition of several other scipy distributions.
 This can be used to combine several components with fixed relative fractions into one pdf, or to create for instance a triple-gaussian distribution (which is used to model gaussian noise coming from different sources).

I attached a minimal example code (as .txt because github doesn't allow .py), which 
uses all 4 distributions.
[pr_example.txt](https://github.com/scipy/scipy/files/601075/pr_example.txt)

Additional Remark:
Usually scipy.stats only adds instances of distributions to its namespace (e.g. norm, but not norm_gen). Since template_gen and mixture_gen are intended to be used by the user directly, I added them to the scipy.stats namespace by hand.
I'm not sure how this interacts with the documentation scipy builds for all distributions automatically,
I appreciate comments on this (and other) matters.

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @thomaskeck, interesting classes. Note that ``template_gen`` has a lot of overlap with gh-6466. It would be good to compare the implementations.

**Comment by rgommers** [comment]
> However the usage / usefulness is not restricted to HEP.

Do you have some examples of this for Argus and Crystal Ball?

**Comment by thomaskeck** [comment]
The remark about the usefulness outside of HEP refers to template_gen and mixture_gen.
To be honest I don't have enough inside in other fields to think of a valid use-case for 1. and 2.
However, as mentioned above the crystalball function describes basically a gaussian distribution with a heavy tail on one side. I imagine there are use-cases for which these kind of distributions are a valid model / parametrization.

The rv_arbitrary looks interesting and would be (once finished) more general than template_gen.
I originally thought about something similar, but sticked to the simple solution in the end.

I will wait for some more comments from you (and others) and remove the distributions from the pull requests which are considered to be too specific.

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
...
+        # Set support
+        kwargs['a'] = 0.0
+        kwargs['b'] = 1.0
+        super(argus_gen, self).__init__(*args, **kwargs)
+
+    def _pdf(self, x, chi):
+        """
+        Return PDF of the argus function
+        """
+        return chi**3 / (_norm_pdf_C * _argus_phi(chi)) * x * np.sqrt(1.0 - x**2) * np.exp(- 0.5 * chi**2 * (1.0 - x**2) )
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Please avoid lines > 80 chars.
E.g. define `y = 1.0 - x**2`, then `return ... x * y * np.exp(-chi**2 * y / 2)`.

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
...
+
+    def _cdf(self, x, chi):
+        """
+        Return CDF of the argus function
+        """
+        return 1.0 - _argus_phi(chi * np.sqrt(1 - x**2)) / _argus_phi(chi)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Better define `_sf` instead, then `_cdf` is automatically `1 - _sf`.

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
...
+argus = argus_gen(name='argus', longname="An Argus Function")
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: need to set `a` and `b` here. Also `_argcheck` to enforce `c > 0`, `chi > 0`.

File: scipy/stats/_continuous_distns.py [hunk 4]
```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
...
+    %(example)s
+    """
+    def __init__(self, *args, **kwargs):
+        """
+        Set finite support for argus function.
+        """
+        # Set support
+        kwargs['a'] = 0.0
+        kwargs['b'] = 1.0
+        super(argus_gen, self).__init__(*args, **kwargs)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: ISTM there is no need to define `__init__` explicitly.

File: scipy/stats/__init__.py [hunk 5]
```diff
@@ -105,6 +108,7 @@
    semicircular      -- Semicircular
    skewnorm          -- Skew normal
    t                 -- Student's T
+   template_gen      -- Template (Generator)
```
Comment (ev-br) [inline scipy/stats/__init__.py]: If we add this and the other one as public functionality, we should find some better names. `_gen` suffixes have been an implementation details and I guess they better stay like this. Also `template_gen` is too generic and non-descriptive.

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
...
+    The pdf and cdf are defined as stepwise functions from the provided histogram.
+    In particular the cdf is not interpolated between bin boundaries and not differentiable.
+
+    %(after_notes)s
+
+    %(example)s
+
+    data = scipy.stats.norm.rvs(size=100000, loc=0, scale=1.5)
+    hist = np.histogram(data, bins=100)
+    template = scipy_extra.stats.template_gen(hist)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: ?

File: scipy/stats/_continuous_distns.py [hunk 7]
```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
...
+    Generates a distribution given by a histogram.
+    This is useful to generate a template distribution from a binned datasample.
+
+    %(before_notes)s
+
+    Notes
+    -----
+    There are no additional shape parameters except for the loc and scale.
+    The pdf and cdf are defined as stepwise functions from the provided histogram.
+    In particular the cdf is not interpolated between bin boundaries and not differentiable.
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: If pdf is piecewise constant, cdf must be piecewise linear. See also a worked example in https://github.com/scipy/scipy/pull/6466#issuecomment-254187399
Feel free to reuse the implementation from the gist linked there.

File: scipy/stats/_continuous_distns.py [hunk 8]
```diff
@@ -5206,6 +5206,321 @@ def _entropy(self, beta):
...
+        pdf = pdf / float(np.sum(pdf * bin_widths)) 
+        cdf = np.cumsum(pdf * bin_widths)[:-1]
+        self.template_bins = bins
+        self.template_bin_widths = bin_widths
+        self.template_pdf = np.hstack([0.0, pdf, 0.0])
+        self.template_cdf = np.hstack([0.0, cdf, 1.0, 1.0])
+        # Set support
+        epsilon = 1e-7
+        kwargs['a'] = self.template_bins[0] - epsilon
+        kwargs['b'] = self.template_bins[-1] + epsilon
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: This `epsilon` is confusing. I think it's better to define the distribution on a closed interval.

File: scipy/stats/tests/test_distributions.py [hunk 9]
```diff
@@ -3008,5 +3008,205 @@ def test_burr12_ppf_small_arg():
...
+        assert_almost_equal(self.template.cdf(5.0), 15.0/25.0)
+        assert_almost_equal(self.template.cdf(6.0), 19.0/25.0)
+        assert_almost_equal(self.template.cdf(7.0), 22.0/25.0)
+        assert_almost_equal(self.template.cdf(8.0), 25.0/25.0)
+        assert_almost_equal(self.template.cdf(9.0), 25.0/25.0)
+        assert_almost_equal(self.template.cdf(10.0), 25.0/25.0)
+
+    def test_rvs(self):
+        N = 10000
+        sample = self.template.rvs(size=N)
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py]: Please always seed the generator in tests. Use `.rvs(..., random_state=123)` or equivalent.

File: scipy/stats/_continuous_distns.py (hunk 7 above)
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Also if this is added in some form, an instance must be added to the test loop in `test_continuous_basic.py`

File: scipy/stats/tests/test_distributions.py [hunk 10]
```diff
@@ -3008,5 +3008,205 @@ def test_burr12_ppf_small_arg():
...
+
+    # for(float x = -5.0; x < 5.0; x+=0.5) std::cout << ROOT::Math::crystalball_cdf(x, 2.0, 3.0, 1.0) << ", ";
+    calculated = stats.crystalball.cdf(X, alpha=2.0, n=3.0)
+    expected = np.array([0.00442081, 0.00559509, 0.00730787, 0.00994682, 0.0143234, 0.0223803, 0.0397873, 0.0830763, 0.173323, 0.320592, 0.508717, 0.696841, 0.844111, 0.934357, 0.977646, 0.993899, 0.998674, 0.999771, 0.999969, 0.999997])
+    assert_allclose(expected, calculated, rtol=0.001)
+
+    # for(float x = -5.0; x < 5.0; x+=0.5) std::cout << ROOT::Math::crystalball_cdf(x, 2.0, 3.0, 2.0, 0.5) << ", ";
+    calculated = stats.crystalball.cdf(X, alpha=2.0, n=3.0, loc=0.5, scale=2.0)
+    expected = np.array([0.0176832, 0.0223803, 0.0292315, 0.0397873, 0.0567945, 0.0830763, 0.121242, 0.173323, 0.24011, 0.320592, 0.411731, 0.508717, 0.605702, 0.696841, 0.777324, 0.844111, 0.896192, 0.934357, 0.960639, 0.977646])
+    assert_allclose(expected, calculated, rtol=0.001)
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py]: Nice comparisons!
Please make all lines < 80 chars however (e.g. implicit line continuation within brackets).

**Review by ev-br — CHANGES_REQUESTED** [review]
This PR tries to add four different things with varying degrees of being close to mergeable. Consider splitting the PR into several ones to keep the scope manageable.

Briefly:
- crystal ball looks useful: a Gaussian + a power law tail, looks useful. You might want to test your implementation a bit in the far tails, to see if you need to define `_sf` in addition to `_cdf`.
- argus needs a bit more work implementation wise. Also  you might want to argue a bit what are the uses of it. E.g., is the plan to provide an incentive to switch away from ROOT?
- template_gen needs a rework (an a rename)
- mixtures is probably a nice feature. However you might want to check the prior art in statsmodels and coordinate with statsmodels devs. I'd ask @josef-pkt for an opinion.

**Comment by ev-br** [comment]
> I attached a minimal example code (as .txt because github doesn't allow .py), which
uses all 4 distributions.
pr_example.txt

This could be better as a gist --- bonus points if you turn it into an IPython notebook, so that the plots are rendered right away.

**Comment by ev-br** [comment]
Also test failures are real.

**Comment by josef-pkt** [comment]
I didn't look much yet.

argus sounds fine. I didn't find much information about it, the wikipedia page is also a bit "sparse".
(It looks a bit like a truncated half-normal, as a one-parameter family on [0, 1] with standard scale=1.

crystallball: I don't like the name much. Matlab has pareto tail distribution, that adds pareto tails to other distributions or to kernel density central distribution. I think it's useful but for consistency it might need 3 versions, lower tail, upper tail and both tails pareto/power.
AFAIR, the python power packages (names ?) use something similar to estimate the threshold at which the power tail(s) start.

Both of those follow the standard pattern. The other two look like `frozen` distributions based on a very brief look. Using a proper pep-8 class name could be a useful convention to distinguish those distributions from the `xxx_gen` global instance distributions.
So, as @ev-br  recommended, I would split out at least those two into separate PRs.

As mentioned in the other PR for generic distributions, I like a histogram distribution because it can be fast given the piecewise constant or linear behavior.

about the mixture distribution: 
statsmodels.distributions also has a version of it that looks superficially very similar. I don't see a reason not to add it to scipy also. One issue will be to decide on broadcasting versus multivariate distributions, and having unit tests for it. 
AFAIR, the statsmodels version leaves it up to the component distributions and works with multivariate component distributions. The purpose of the statsmodels version was mainly to provide a convenient way to generate mixture samples (rvs), and we didn't put a lot of effort on design and completeness.

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (thomaskeck) [inline scipy/stats/_continuous_distns.py]: As far as I understood rv_continuous automatically shifts and scales the boundaries with respect to loc and scale. So I just set a=0, b=1 at construction time and remove __init__.

File: scipy/stats/_continuous_distns.py (hunk 3 above)
Comment (thomaskeck) [inline scipy/stats/_continuous_distns.py]: rv_continuous default implementation of _argcheck already enforces positive chi.

**Comment by thomaskeck** [comment]
Thanks for the extensive feedback.

I implement the proposed changes, and open new PRs for each of the distributions.

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (thomaskeck) [inline scipy/stats/_continuous_distns.py]: This doesn't seem to be the case.
If I define only _pdf and _sf, _cdf seems to be calculated using _pdf integration.

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: > As far as I understood rv_continuous automatically shifts and scales the boundaries

yes, you're right. To make it absolutely sure, please add a test with scale != 1

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: ÒK, yes indeed it does. Then I'd add `def _cdf(...): return 1.0 - self._sf(...)`.

**Comment by thomaskeck** [comment]
Following your suggestions I splitted up this PR:

Argus https://github.com/scipy/scipy/pull/6799
Crystalball https://github.com/scipy/scipy/pull/6798
Mixture https://github.com/scipy/scipy/pull/6800
Histogram https://github.com/scipy/scipy/pull/6801

I also implemented a first round of corrections, and fixed the unittests in test_continuous_basic.py.
The most important comments for Histogram (there is a more general approach named rv_arbitrary),
and mixture (there is maybe something similar statsmodels) are mentioned in the new PRs.

I close this PR now, so the discussion can be focused in the new PRs.

Best regards,
Thomas

**Pull request closed without merging**